### ISSUE TITLE
Create tooling to take a FCOS-derived container and generate bootable images from it

### ISSUE THREAD

**Issue by cgwalters** [coreos/fedora-coreos-tracker#1151]
This is touched on in https://github.com/coreos/enhancements/blob/main/os/coreos-layering.md#use-of-coreos-diskboot-images

Basically I think our default stance is that admins use our golden images + ignition to target their custom containers, but I think for a variety of reasons we will want to support generating disk images directly too.

I think this would be relatively maintainable for bare metal targets in particular (ISO, PXE etc.), and it's on bare metal targets where there's more of a desire for things like handling the first boot with an out of tree NIC driver, etc.

Now, obviously this is basically what coreos-assembler is today.  However, a few key points:

- We don't want to expose all of coreos-assembler as a supported UX
- The input to this tool should *only* be the derived container image, not e.g. the config git repository too.  This topic heavily touches on https://github.com/coreos/coreos-assembler/issues/2685

This build tool could derive from the coreos-assembler container literally, or we could split out the bits we need as a shared library or so (could make sense as a Go library actually, even if we happen to ship shell scripts and python in it?).  Starting out this way is going to be the most supportable thing.  This tool should be runnable as a container image (that yes, accepts a container image as input) by default, but we could support other flows too.

That said, we should also do a spike on having e.g. [RHEL Image Builder](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/composing_a_customized_rhel_system_image/composer-description_composing-a-customized-rhel-system-image) be able to output an ISO at least.

**Comment by cgwalters** [comment]
I briefly looked at trying to have a new tool that "restamps" an *existing* live ISO with new content by scraping out the files, but what makes this hard right now is that https://github.com/coreos/coreos-assembler/blob/main/src/cmd-buildextend-live has a whole lot of load-bearing magic around things like the kernel arguments embed area and architecture specifics.

Further, that code specifically also reads `src/config/live` which is content *outside* of the container image today.  We'd have to switch to embedding that content inside the target image, similar to what we did with https://github.com/coreos/coreos-assembler/commit/816ebaed49d3e1efb2ba06e52969453ddad06d61

**Comment by cgwalters** [comment]
Been thinking about this more.  How about this plan:

- Take all the logic from https://github.com/coreos/coreos-assembler/blob/main/src/cmd-buildextend-live and move it to coreos-installer in Rust
- It now works to run `podman run quay.io/examplecorp/custom-fcos-derivative:latest coreos-installer generate-iso --output=- > examplecorp-custom.iso` which just uses as source of truth data from *the running container filesystem*
- coreos-assembler just runs this (inside supermin I guess)

**Comment by ybettan** [comment]
Regarding https://github.com/coreos/fedora-coreos-tracker/issues/1151#issuecomment-1351656773:

If we are re-generating the ISO using the coreos-installer then why is it better than using the coreos-assembler for it? I though that the benefit of the installer is not manipulate the ISO without rebuilding it from source each time.

**Comment by cgwalters** [comment]
My latest work on this is in https://github.com/containers/bootc/pull/30

**Comment by cgwalters** [comment]
The initial design of `coreos-installer` was to install raw disk images - with a few very small tweaks like injecting Ignition.  Since then, its role has grown significantly.  But adding container installation to it would IMO completely change the character of the project.

Further, one wants to use *the same* technology stack when fetching updates (aside from writing the partitions etc.).  This is why I created the new bootc project, which is exactly this.

I would like to add bootc to FCOS, and am working in that direction.

**Comment by jmpolom** [comment]
@cgwalters that makes a lot of sense. Do you see bootc also sidestepping the need to add functionality to ignition/butane to configure an installed system to boot a containerized build? Such as what is hinted at in the [coreos layering enhancements](https://github.com/coreos/enhancements/blob/main/os/coreos-layering.md#use-of-coreos-diskboot-images) document.

I'm assuming that what is described at the end of that document is not currently available functionality. I did not see it mentioned in the latest butane documentation.

**Comment by cgwalters** [comment]
>  Do you see bootc also sidestepping the need to add functionality to ignition/butane to configure an installed system to boot

Yes...and no.  Yes, `bootc` allows directly installing customized container images without an intermediate disk image being generated.  But, I think we should still have butane sugar for this to target rebasing an existing FCOS disk image.

**Comment by cgwalters** [comment]
Also though of note, this flow is only helpful for the scenario where one can get a machine up enough to run this code and write to disk.  There are some cases where we can e.g. attach a stock ISO via IPMI or so, then do this installation to disk with the real NIC drivers already embedded etc.  Or, plugging in a USB stick (say the FCOS-derivative Live ISO) but also with a data partition containing the target container image.

It'd also be possible, albeit hacky to e.g. inject ignition into our live ISO which downloads and loads the kmod dynamically.  (Or I guess...assuming the live ISO kernel is the same as the container kernel, could actually extract it from the container)

But ultimately we do need some sort of ergonomic tool which takes a container image as input and spits out an ISO that can install to disk an OS that will consume that modified container image to boot.

Hmmm....I wonder though, maybe short term we could try to make some tool which actually does the "data partition with container" thing more streamlined.  That would help with a few more cases perhaps, but ultimately we really do need something that has the ISO boot reliably from the kernel that's in the container along with any kmods.

**Comment by cgwalters** [comment]
BTW though, there are still a lot of architectural problems with the ["hackily update after coreos-installer"](https://github.com/coreos/fedora-coreos-tracker/issues/1151#issuecomment-1430263412) approach.

For example, because things like LUKS are handled via Ignition always today, it means that the files we write from the container image are unencrypted.  For Fedora CoreOS derivatives, this doesn't matter because everything we ship is FOSS - there is no secret content.

I think that's *not* going to be true in the general case for custom derived images; I'm pretty sure some users will embed secret data into images (passwords or keys).  We can't really say it's a bad idea to do that unless we offer them a significantly more ergnonomic/compelling way to do it.  (This does touch on [bootc configmap/secret support](https://github.com/containers/bootc/issues/22) some, particularly if we do something nice with secrets).  I think Ignition has always had this problem, but Ignition configs also happen on firstboot today and the files it writes will happen *after* LUKS has been enabled.

**Comment by jmpolom** [comment]
I would say as a downstream user/consumer of whatever functionality is developed, a way to directly install a container image to disk explicitly would be my preferred workflow. I have thought about doing the ["post install rebase/deploy hack"](https://github.com/coreos/fedora-coreos-tracker/issues/1151#issuecomment-1430263412) but it really does seem like exactly that and I'm apt to avoid such hacks at any scale.

I think having a simple utility without extensive dependencies that could generate the raw bootable disk image (could this step require qemu virtualization? that would be a drag) that could be consumed by `coreos-installer` as-is would be preferable to me as a user/consumer. I understand not wanting to complicate what should be a simple tool (an installer of raw disk images) that has already evidently experienced scope creep. Expanding the capability and scope of the installer doesn't seem wise.

As for ignition, I think if there's a utility to create disk images and ISOs that can be used by the installer to be written to disk (or booted over the network?!) then there is likely little need for support in ignition. If ignition has to do the rebase or deploy that implies a plain FCOS image is installed, booted, ignition runs and then installs a customized container image. This seems terribly inefficient and convoluted.

Ultimately I think bootc is my true answer here so hopefully that project matures, stabilizes and sees good adoption.

**Comment by cgwalters** [comment]
Just to xref, there is a new "more incremental" potential path here https://issues.redhat.com/browse/OCPBU-66?focusedId=22251336&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-22251336 that's being discussed in the OCP context but would apply to FCOS too

**Comment by cgwalters** [comment]
Use `$ignition_firstboot` instead of `ignition.firstboot`

**Comment by cgwalters** [comment]
For doing OS updates at coreos-installer time, the topic of kernel arguments came up (both Ignition and MachineConfig for OCP).  Basically we need to extract the kernel arguments from the Ignition/MC and apply them at this time too.

**Comment by ori-amizur** [comment]
@cgwalters I implemented https://github.com/coreos/fedora-coreos-tracker/issues/1151#issuecomment-1430263412 and it works, except that when I moved to 4.14  image I got errors from selinux:
```
error: Performing deployment: Importing: Layer sha256:dc1def201e9ad26de700cc96e92dad71ebfb917de1e8155f9affe2d0abd1fcd0: Importing object bd/862ae43b9c5f71e7513e91b71b419b2436a556faaa59204441094856dcc493.file: Processing content object bd862ae43b9c5f71e7513e91b71b419b2436a556faaa59204441094856dcc493: Setting xattrs: fsetxattr(security.selinux): Invalid argument
```
When I disabled selinux (**setenfoce 0**), the ostree command was successful.
Any idea how to overcome the problem?

**Comment by cgwalters** [comment]
> When I disabled selinux (setenfoce 0), the ostree command was successful.

Yes...it's possible but tricky to handle this.  We do it in bootc today, see https://github.com/containers/bootc/blob/0a657ea9609e33c09e8bf2c013699c379f0ccc43/lib/src/lsm.rs#L32

But `setenforce 0` is indeed the most reliable way.  It's how Anaconda has always worked.

**Comment by Smithx10** [comment]
@cgwalters,

I am hoping to build a pxe image with zfs added.  My intention is to run the os from ram only and never install it onto disk.

I think it would be helpful to have an example that uses https://github.com/coreos/layering-examples/tree/main/build-zfs-module to demonstrate creating a bootable image that has zfs installed.  

Has this issued progressed far enough to do this now?   If so could you please show me how I would build one?  

Thanks for all your effort solving this problem.

**Comment by jlebon** [comment]
We now have https://github.com/coreos/custom-coreos-disk-images, which is more of a POC repo to show that it is possible and it's certainly usable today. But we don't expect to document it more broadly for now.

The end goal is definitely to have common tooling for building disk images (IMO both for _official_ artifacts and users' custom artifacts), whether that ends up being [bootc-image-builder](https://github.com/osbuild/bootc-image-builder) or something else. That is also tracked as part of https://github.com/coreos/fedora-coreos-tracker/issues/1726.
